src/component/mod.rs

```diff
@@ -164,6 +164,18 @@ impl ComponentManager {
             active_components,
         }
     }
+
+    // The heartbeat function is called periodically
+    pub fn heartbeat(&self) {
+        for component in self.active_components.values() {
+            // It's okay not to block on the lock -- heartbeats have no guaranteed periodicity
+            // (Plus, this is only used for component shutdown, if someone has this lock, the
```

Comment:
While this is true in theory, have you guaranteed that there can't be a situation where an empty/shutdown process isn't holding any locks?

Reply:
I don’t follow the question. In order for a lock to be taken, some operation is happening on the component. It’s okay if we don’t ‘hearbeat’ an in use component

Reply:
I don’t follow the question. In order for a lock to be taken, some operation is happening on the component. It’s okay if we don’t ‘hearbeat’ an in use component